client/trino-jdbc/src/main/java/io/trino/jdbc/TrinoConnection.java

```diff
@@ -156,6 +171,74 @@ public class TrinoConnection
 
         uri.getExplicitPrepare().ifPresent(value -> this.useExplicitPrepare = value);
         uri.getAssumeNullCatalogMeansCurrentCatalog().ifPresent(value -> this.assumeNullCatalogMeansCurrentCatalog = value);
+
+        validateConnection = uri.isValidateConnection();
+        if (validateConnection) {
+            try {
+                if (!validateCredentials(CONNECTION_TIMEOUT_SECONDS)) {
+                    throw new SQLException("Invalid authorization specification.", "28000");
+                }
+            }
+            catch (UnsupportedOperationException | IOException e) {
+                throw new SQLException("Unable to connect.", "08001", e);
+            }
+        }
+    }
+
+    private boolean validateCredentials(int timeout)
+            throws IOException
+    {
+        HttpUrl url = HttpUrl.get(httpUri);
+        url = url.newBuilder().encodedPath("/v1/statement").build();
+
+        Request.Builder builder = new Request.Builder()
+                .url(url)
+                .head();
+        Request request = builder.build();
+
+        Exception lastException = null;
+        Duration timeoutDuration = new Duration(timeout, TimeUnit.SECONDS);
+        long start = System.nanoTime();
+
+        while (true) {
+            Duration sinceStart = Duration.nanosSince(start);
+            if (timeout > 0 && sinceStart.compareTo(timeoutDuration) > 0) {
+                throw new IOException(format("ValidateCredentials timed out after %s. Last error: %s", sinceStart,
+                        lastException == null ? "null" : lastException.getClass().getName()), lastException);
+            }
+
+            try (Response response = httpCallFactory.newCall(request).execute()) {
+                switch (response.code()) {
+                    case HTTP_OK:
+                        return true;
+                    case HTTP_UNAUTHORIZED:
+                        return false;
+                    case HTTP_BAD_METHOD:
+                        throw new UnsupportedOperationException("Target server does not support HEAD /v1/statement");
```

Comment:
`return true` for backward compatibility

Reply:
`TrinoConnection.validateCredentials(int)` is called by both `TrinoConnection.<init>` and `TrinoConnection.isValid(int)`, and only when `validateConnection` is explicitly set, so there is no backward compatibility issue here.

In order to distinguish the following three cases for `TrinoConnection.<init>`, this method need to return `true` (everything is fine), `false` (invalid authorization) and `exception` (cannot connect caused by other errors).

Reply:
I'd prefer this to be always on and drop the validateConnection property